initiate new handshake after restart of receiving system, #20568

patriknw opened the pull request:
- we don't want to include the full origin address in each message,
  only the UID
- that means that the restarted receiving system can't initate a
  new handshake immediately when it sees message from unknown origin
- instead we inject HandshakeReq from the sending system once in a while
  (1 per second) which will trigger the new handshake
- any messages that arrives before the HandshakeReq are dropped, but
  that is fine since the system was just restarted anyway
- note that the injected handshake is only done for active connections,
  when a message is sent
- also changed the UID to a Long, but there are more places in old remoting
  that must be changed before we actually can use a Long value

patriknw committed:
initiate new handshake after restart of receiving system, #20568

* we don't want to include the full origin address in each message,
  only the UID
* that means that the restarted receiving system can't initate a
  new handshake immediately when it sees message from unknown origin
* instead we inject HandshakeReq from the sending system once in a while
  (1 per second) which will trigger the new handshake
* any messages that arrives before the HandshakeReq are dropped, but
  that is fine since the system was just restarted anyway
* note that the injected handshake is only done for active connections,
  when a message is sent
* also changed the UID to a Long, but there are more places in old remoting
  that must be changed before we actually can use a Long value

fix lost first message, #20566

* the first message was sometimes dropped by the InboundHandshake stage
  because it came from unknown origin, i.e. the handshake had not completed
* that happended because the ordinary messagage arrived before the
  first HandshakeReq, which may happen since we sent the HandshakeReq
  over the control stream
* this changes so that HandshakeReq is sent over the same stream, not
  only on the control stream and thereby the HandshakeReq will arrive
  before any other message
* always send HandshakeReq as first message
  * also when the handshake on sender side has been completed at startup
  * moved code from preStart to onPull

patriknw commented:
Refs #20568

patriknw commented:
I found the mystery with the lost first message, see description in commit 1294b09

patriknw commented:
Refs #20566

drewhk commented on akka-remote/src/main/scala/akka/remote/AddressUidExtension.scala:
```diff
@@ -25,5 +25,12 @@ object AddressUidExtension extends ExtensionId[AddressUidExtension] with Extensi
-}
\ No newline at end of file
+  val longAddressUid: Long = {
+    // FIXME we should use a long here, but then we need to change in Cluster and RemoteWatcher also
```
We might keep them as-is, we can give those only the low 32 bits. But it is better to use all 64 bits of course if possible.

patriknw replied:
My point is that we can't use different uid for cluster and remoting, becuase that uid is used by clustering when it quarantines a removed node and then it must be the full length uid if that is what is used in remoting. But it might be possible that we use a full length uid under the hood in clustering, but only expose the low 32 for users (in existing api, such as `Member`)

patriknw replied:
I have created a separate ticket for this: https://github.com/akka/akka/issues/20644

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -103,6 +103,16 @@ private[akka] trait InboundContext {
+  /**
+   * Lookup the outbound association for a given UID.
+   * Will return `null` if the UID is unknown, i.e.
+   * handshake not completed. `null` is used instead of `Optional`
```
we really need a value class for these, mimicking Option.

ktoso replied:
👍 Yeah, might be a good time to roll one during artery work finally.

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -150,7 +160,7 @@ private[akka] final class AssociationState(
 
   def isQuarantined(uid: Long): Boolean = {
-    // FIXME does this mean boxing (allocation) because of Set[Long]? Use specialized Set. LongMap?
+    // FIXME does this mean boxing (allocation) because of Set[Long]? Use specialized Set. org.agrona.collections.LongHashSet?
```
+1 for the agrona set. I wish they would use a better hash function though.

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -253,7 +262,8 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
   // FIXME: This does locking on putIfAbsent, we need something smarter
-  private[this] val associations = new ConcurrentHashMap[Address, Association]()
+  private[this] val associationsByAddress = new ConcurrentHashMap[Address, Association]()
```
I think it is time to move this to a class in itself, so it can make sure its invariants are kept intact (updating the two map properly). I also feel a risk of races here, accessing the two maps concurrently.

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -476,15 +486,28 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
-  override def quarantine(remoteAddress: Address, uid: Option[Int]): Unit =
-    association(remoteAddress).quarantine(reason = "", uid) // FIXME change the method signature (old remoting) to include reason?
+  override def quarantine(remoteAddress: Address, uid: Option[Int]): Unit = {
+    // FIXME change the method signature (old remoting) to include reason and use Long uid?
```
yes, I think we should.

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/Handshake.scala:
```diff
@@ -59,50 +59,74 @@ private[akka] class OutboundHandshake(outboundContext: OutboundContext, timeout:
+
+        // inject a HandshakeReq once in a while to trigger a new handshake when destination
+        // system has been restarted
+        // FIXME if nanoTime for each message is too costly we could let the TaskRunner update
```
Can't we just use a normal timer here? The injection interval is not too frequent so that should be fine I think.

patriknw replied:
I guess I was thinking that it should not tick when no messages are sent, but that is solvable anyway. Good suggestion, I'll change.

patriknw commented:
@drewhk I rewrote the injection using a timer instead, that's better, thanks. ae769b6
Also extracted the maps into `AssociationRegistry` class. dcd2c4c

patriknw commented:
rebased

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -48,14 +48,14 @@ private[remote] object EnvelopeBuffer {
+  val ClassManifestTagOffset = 24 // Int
 
-  val LiteralsSectionOffset = 32
+  val LiteralsSectionOffset = 28
```
@drewhk did you leave some extra space in front of the LiteralsSectionOffset on purpose? perhaps nice to start it on full byte?

drewhk replied:
No, I think I just shifted stuff once and forgot to move that one. The only thing that can matter is 4 byte allignment. We might be able to pack things even more.

patriknw replied:
ok, thanks for clarifying
